pymidi/server.py

```diff
@@ -74,31 +80,49 @@ def _midi_command_cb(self, peer, midi_packet):
         for handler in self.handlers:
             handler.on_midi_commands(peer, commands)
 
-    def serve_forever(self):
-        control_port, data_port = self.port, self.port + 1
-        logger.info('Control socket on {}:{}'.format(self.host, control_port))
-        control_socket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
-        control_socket.bind((self.host, control_port))
-
-        logger.info('Data socket on {}:{}'.format(self.host, data_port))
-        data_socket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
-        data_socket.bind((self.host, data_port))
-
-        data_protocol = DataProtocol(data_socket, midi_command_cb=self._midi_command_cb)
-        control_protocol = ControlProtocol(
-            data_protocol,
-            control_socket,
+    def _build_control_protocol(self, host):
+        family = socket.AF_INET
+        if '::' in host:
+            family = socket.AF_INET6
+
+        logger.info('Control socket on {}:{}'.format(host, self.port))
+        control_socket = socket.socket(family, socket.SOCK_DGRAM)
+        control_socket.bind((host, self.port))
+        return ControlProtocol(
+            socket=control_socket,
             connect_cb=self._peer_connected_cb,
             disconnect_cb=self._peer_disconnected_cb)
 
+    def _build_data_protocol(self, host):
+        family = socket.AF_INET
+        if '::' in host:
+            family = socket.AF_INET6
+
+        logger.info('Data socket on {}:{}'.format(host, self.port+1))
+        data_socket = socket.socket(family, socket.SOCK_DGRAM)
+        data_socket.bind((host, self.port+1))
+        return DataProtocol(data_socket, midi_command_cb=self._midi_command_cb)
+
+    def _init_protocols(self):
+        for host in self.host, self.ipv6_host:
```

Comment:
You might want to do something like:

```py
for host, family in ((self.host, socket.AF_INET), (self.ipv6_host, socket.AF_INET6)):
  data_protocol = self._build_data_protocol(host, family)
  ctrl_protocol = ...
```

..rather than infer the family from the value. The reason I am thinking of is that _technically_, you can bind to a hostname (like `localhost`), which could resolve to a v4 or v6 addr & would defeat the family test you have on L85.

Reply:
Yah, that makes sense.  I wasn't super happy with inferring the family, and started looking at holding the host in an ipaddress object.  However I didn't feel like adding another dependency just for that and I wanted to get something working.

Updated!